src/pgwire/src/message.rs

```diff
@@ -796,6 +885,142 @@ pub fn decode_copy_text_format(
     Ok(rows)
 }
 
+pub struct CopyCsvFormatParams<'a> {
+    delimiter: u8,
+    quote: u8,
+    escape: u8,
+    header: bool,
+    null: Cow<'a, str>,
+}
+
+impl<'a> TryFrom<CopyParams> for CopyCsvFormatParams<'a> {
+    type Error = ErrorResponse;
+
+    fn try_from(
+        CopyParams {
+            format,
+            null,
+            delimiter,
+            quote,
+            escape,
+            header,
+        }: CopyParams,
+    ) -> Result<Self, Self::Error> {
+        assert_eq!(format, CopyFormat::Csv);
+
+        fn extract_byte_param_value(
+            v: Option<String>,
+            default: u8,
+            param_name: &str,
+        ) -> Result<u8, ErrorResponse> {
+            Ok(match v {
+                Some(v) if v.len() == 1 => v.as_bytes()[0],
+                Some(..) => {
+                    return Err(ErrorResponse::error(
+                        SqlState::FEATURE_NOT_SUPPORTED,
+                        format!("COPY {} must be a single one-byte character", param_name),
+                    ))
+                }
+                None => default,
+            })
+        }
+
+        let null = match null {
+            Some(null) => Cow::from(null),
+            None => Cow::from(""),
+        };
+        let delimiter = extract_byte_param_value(delimiter, b',', "delimiter")?;
+        let quote = extract_byte_param_value(quote, b'"', "quote")?;
+        let escape = extract_byte_param_value(escape, quote, "escape")?;
+        let header = header.unwrap_or(false);
+
+        Ok(CopyCsvFormatParams {
+            delimiter,
+            quote,
+            escape,
+            null,
+            header,
+        })
+    }
+}
+
+pub fn decode_copy_format_csv(
+    data: &[u8],
+    column_types: &[pgrepr::Type],
+    CopyCsvFormatParams {
+        delimiter,
+        quote,
+        escape,
+        null,
+        header,
+    }: CopyCsvFormatParams,
+) -> Result<Vec<Row>, io::Error> {
+    let mut rows = Vec::new();
+
+    let (double_quote, escape) = if quote == escape {
+        (true, None)
+    } else {
+        (false, Some(escape))
+    };
+
+    let mut rdr = ReaderBuilder::new()
+        .delimiter(delimiter)
+        .quote(quote)
+        .has_headers(header)
+        .double_quote(double_quote)
+        .escape(escape)
+        // Must be flexible to accept end of copy marker, which will always be 1
+        // field.
+        .flexible(true)
+        .from_reader(data);
+
+    let records = rdr
```

Comment:
Is there a reason to pre-parse the whole thing instead of row-by-row?

Reply:
Just an oversight; fixed